Library/OcCpuLib/OcCpuLib.c

```diff
@@ -661,46 +663,70 @@ ScanIntelProcessor (
   // SkyLake and later have an Always Running Timer
   //
   if (Cpu->Model >= CPU_MODEL_SKYLAKE) {
-    AsmCpuid (CPUID_TIME_STAMP_COUNTER, &CpuidEax, &CpuidEbx, NULL, NULL);
+    AsmCpuid (CPUID_TIME_STAMP_COUNTER, &CpuidEax, &CpuidEbx, &CpuidEcx, NULL);
+    if (CpuidEcx > 0) {
+      Cpu->ARTFrequency = CpuidEcx;
+      DEBUG ((DEBUG_INFO, "OCCPU: Core Crystal Clock Frequency %u\n", CpuidEcx));
+    } else {
+      //
+      // Fall back to identifying ART frequency based on model
+      //
+      if (Cpu->Family == 0x6 && Cpu->Model == CPU_MODEL_SKYLAKE_W) {
+        //
+        // TODO: Xeon Ws share the same CPUID but I believe they have 24 Mhz ART
+        //
+        Cpu->ARTFrequency = 25000000ULL; // 25 Mhz
```

Comment:
Yes, I can confirm that Xeon W, Xeon X, and Xeon Scalable have all the same Platform ID and CPUID but different frequency. This is terrible, and I am not sure how we can handle it. As a last resort option we can use CPUID brand string, but I hope that something better arises. I will check the docs when I get a spare minute.

Reply:
For now, I adjusted `DetectAppleMajorType` to return a new `enum` value `AppleProcessorMajorXeonScalable` (determined from the brand string) which we can use to differentiate. I believe that `enum` is for internal use only anyways so that should be fine. I pun the `AppleProcessorMajorXeonScalable` into a Xeon W in `DetectAppleProcessorType` now.

I don't think it makes the code much uglier, thankfully.

Reply:
Hmm, issue is, `AppleProcessorMajorXeonW` is not internal, it is Apple value actually, and they do not have Xeon Scalable. Thus the reason it is in a separate EfiPkg repository. If anything, we should at least name it differently.

I need to really think if there is any better way to distinguish the CPUs. Currently I dislike the fact to need to check the brand string, though it will work, yes.

Reply:
Alright, I rolled that back. I've tacked it onto `DetectAppleMajorType` as a `BOOLEAN` additional `OUT` argument called `IsXeonScalable` until we can come up with a more permanent or better solution.